### ISSUE TITLE
Allow bypassing auto-detection of function getrandom (for both Autotools and CMake build systems)

### ISSUE THREAD

**Issue by piponazo** [libexpat/libexpat#239]
Hi!

I would like to report an issue I am having while trying to update the Expat dependency in the Exiv2 project, from the version 2.2.5 to the version 2.2.6. 

In Exiv2 we bring the expat dependency (and others) with conan. I actually wrote a conan recipe for expat which is maintained here:
https://github.com/Pix4D/conan-expat

As you can see here, in this Exiv2 branch I am just trying to bump the expat version from 2.2.5 to 2.2.6:
https://github.com/Exiv2/exiv2/compare/master...piponazo:expat226

And by just doing that change, one of our travis CI jobs started to fail with this error message:
```
[ 52%] Linking CXX executable ../bin/exiv2
../lib/libexiv2.so.0.27.0.2: undefined reference to `getrandom'
```
This is just happening for the **Linux job (Ubuntu 14.04) with clang-5**. The rest of jobs in the matrix compile the project correctly:
https://travis-ci.org/piponazo/exiv2/builds/458033518

I saw that there are other issues related somehow with the `getrandom` function (for example #172). I would like to know if I need to add some definition (and which one) to solve this issue. 

Thanks in advance! 😸

**Comment by hartwork** [comment]
Hi!

This reminds me of #238. If there's an extra layer on top of Expat — conan in this case — please either consult with your "integrator" — the person packaging Expat for conan. If that turns out to point upstream, I'm happy to join the chase down the rabbit hole.

On a side note, `getrandom` requires glibc >=2.25. Please check if that's what you're linking against.

Best, Sebastian

**Comment by piponazo** [comment]
I am the person who wrote that conan recipe for expat, so I know the details about it :wink: .

In conan I use CMake to configure the expat project just with this snippet of code:

```python
def source(self):
        self.run("git clone --depth 1 --branch R_2_2_6 %s" % self.source_url)

    def build(self):
        cmake = CMake(self)

        cmake_args = { "BUILD_doc" : "OFF",
                       "BUILD_examples" : "OFF",
                       "BUILD_shared" : self.options.shared,
                       "BUILD_tests" : "OFF",
                       "BUILD_tools" : "OFF",
                       "CMAKE_POSITION_INDEPENDENT_CODE": "ON",
                       "MSVC_USE_STATIC_CRT": self.options.static_crt,
                     }

        cmake.configure(source_dir="../libexpat/expat", build_dir="build", defs=cmake_args)
        cmake.build()
        cmake.install()
```
The conan packages are generated in TravisCI for Linux and OSX for a big combination of compilers (gcc-4, gcc-5, gcc-6, gcc-7, gcc-8, clang-4, clang-5, clang-6) and options (Release, Debug, Shared, Static, etc). You can see a specific example of the compilation for Linux + clang-5 in this [link](https://travis-ci.org/piponazo/conan-expat/jobs/456980455). Note that the images used in travis for Linux are Ubuntu 14.04. The glibc version might be older than 2.25. However, in case that would be the reason of the failure, I would expect to see all my CI jobs failing, and not only the one for clang-5. You can see here that exiv2 is compiling correctly for other compilers:
https://travis-ci.org/piponazo/exiv2/builds/458033518

Right now, that conan-recipe is checking out the libexpat tag **R_2_2_6** as you can see in the previous python snippet. Do you know if there has been any relevant change related to this `getrandom()` function between the 2.2.6 release and master?

**Comment by hartwork** [comment]
PS: Part of 6b3c44815396d4464c6e5e228eb281a9d206ca54 is this change:
```
# git diff R_2_2_5 R_2_2_6 -- ConfigureChecks.cmake | fgrep getrandom
-check_function_exists("getrandom" HAVE_GETRANDOM)
+check_symbol_exists("getrandom" "sys/random.h" HAVE_GETRANDOM)
```
@bradking have a minute? It looks like we do need to link for `getrandom` to avoid false positives. What do you think?

**Comment by piponazo** [comment]
Humm ... I suspect that the problem might be that when the conan packages are generated some docker images prepared for each compiler are used. This is how the **.travis.yml** file looks for a conan recipe:

```
matrix:
   include:

      - <<: *linux
        env: CONAN_GCC_VERSIONS=4.9 CONAN_DOCKER_IMAGE=lasote/conangcc49

      - <<: *linux
        env: CONAN_GCC_VERSIONS=5 CONAN_DOCKER_IMAGE=lasote/conangcc5

      - <<: *linux
        env: CONAN_GCC_VERSIONS=6 CONAN_DOCKER_IMAGE=lasote/conangcc6

      - <<: *linux
        env: CONAN_GCC_VERSIONS=7 CONAN_DOCKER_IMAGE=lasote/conangcc7

      - <<: *linux
        env: CONAN_GCC_VERSIONS=8 CONAN_DOCKER_IMAGE=lasote/conangcc8

      - <<: *linux
        env: CONAN_CLANG_VERSIONS=3.9 CONAN_DOCKER_IMAGE=lasote/conanclang39

      - <<: *linux
        env: CONAN_CLANG_VERSIONS=4.0 CONAN_DOCKER_IMAGE=lasote/conanclang40

      - <<: *linux
        env: CONAN_CLANG_VERSIONS=5.0 CONAN_DOCKER_IMAGE=lasote/conanclang50

      - <<: *linux
        env: CONAN_CLANG_VERSIONS=6.0 CONAN_DOCKER_IMAGE=lasote/conanclang60
...
```

In the dockerfiles for each of those images, I do not see anything related with **glibc** but maybe it is getting explicitly updated? As an example, this the dockerfile for [conangcc8](https://github.com/conan-io/conan-docker-tools/blob/master/gcc_8/Dockerfile).

In conclusion, my guess is that when I generate the conan packages for expat we might have a more recent version of **glibc** with getrandom, and when compiling exiv2 (with normal travis images) we do not have the same version of **glibc**, but the expat package we are fetching from the conan-servers have been compiled assuming that `getrandom` was awailable. Does it make sense for you?

In case my assumption is right ... is there any way to disable the usage of `getrandom` at the CMake configuration time ? In case such option exists, I think it would be the easier way to fix the situation for my use case.

**Comment by piponazo** [comment]
I think I was right in the previous analysis. I kept thinking about this and then I realised that this situation is only happening in the CI jobs for the Exiv2 project, which are using the default Linux Images:

https://travis-ci.org/piponazo/exiv2/builds/458033518

I also wrote a conan recipe for **exiv2** with was using already expat 2.2.6 and there I did not suffer this problem:
https://travis-ci.org/piponazo/conan-exiv2 

So, why am I having this issue in the exiv2 CI jobs, but not in the conan-exiv2 ones? I think the reason is that the conan-exiv2 job are also using the docker images provided by conan, which probably update the glibc version. 

The issue is a bit obscure and difficult to analyse. From my point of view, I think it would be interesting to have an option in CMake (and probably in the configure script) to disable the usage of `getrandom` even if it is available in the system where expat is being compiled. If you deploy such binaries into a system with an old glibc version, you'll obtain these linking issues. 

What do you think?

Cheers

**Comment by hartwork** [comment]
> From my point of view, I think it would be interesting to have an option in CMake (and probably in the configure script) to disable the usage of getrandom even if it is available in the system where expat is being compiled.

Would `sed 's,.*HAVE_GETRANDOM.*,/* #undef HAVE_GETRANDOM */,' -i expat_config.h` after `./configure` / `cmake .` work for you?

If it does not, any chance you could help out with a pull request?

**Comment by piponazo** [comment]
I think it would make sense. I'll try to add that in the conan expat recipe and I'll give you feedback about it.

**Comment by piponazo** [comment]
I have implemented the change you proposed in the conan-expat recipe:
https://github.com/Pix4D/conan-expat/pull/11/files

And later I have tested the packages in the exiv2 project:
https://travis-ci.org/piponazo/exiv2/builds/458818352

Now everything works like a charm :smile: .

I now that this might look like a very corner case, but as you commented, in #238 it seems they had a similar issue. I think it would be interesting to have an option in your configure and CMake scripts to provide users the option of disabling the usage of `getrandom`. Or at least, it would be good to document this issue somewhere.

Anyways, since now I understood the issue and I could deal with it without touching expat, I let you decide if you want to close this issue now.

Thanks for the help!

**Comment by hartwork** [comment]
Thanks for the update. I'll turn this ticket into a feature request and keep it open in case someone wants to contribute a pull request later.

So what's missing is an option for `CMakeLists.txt` and `configure.ac` to bypass auto-detection of `getrandom`.

**Comment by MohammedKhajapasha** [comment]
hartwork, is this issue fixed ? looks the change change has been merged in canon-expat package.

**Comment by hartwork** [comment]
The conan approach might work for conan, but not for everyone, I'm afraid.

**Comment by kkkunche** [comment]
hartwork, could you please review the pull request https://github.com/libexpat/libexpat/pull/277.

I have added a option in CMake, to disable auto detect of get random.
$mkdir buiild
$cd build
$cmake .. -DHAVE_GETRANDOM=ON/OFF

I will do a similar change for autotools.

**Comment by kkkunche** [comment]
Under the same pull request we have another commit for Autotools

**Comment by hartwork** [comment]
CMake aspect covered by #277 now. Autotools left to do.

**Comment by kkkunche** [comment]
Working on Autotools

**Comment by kkkunche** [comment]
Could you please review the changes for Autotools
https://github.com/libexpat/libexpat/pull/289/commits/bbbf256d82623512820e4f7124304acf8a160993

**Comment by hartwork** [comment]
> Could you please review the changes for Autotools
> [bbbf256](https://github.com/libexpat/libexpat/commit/bbbf256d82623512820e4f7124304acf8a160993)
> 
> Please let me know if any changes needed.

Please use the "Request review" button at #289 instead of writing here, if I don't provide feedback in time. I will need at least half a day more this time. Thanks!

**Comment by kkkunche** [comment]
Thanks and will definitely follow that.

**Comment by hartwork** [comment]
Just merged in aa6d76e9d947b3f09e4ac4adca22deef56ae7712, closing.
